Introduce a CompactVariant for efficient variants of pointers and types smaller than 56 bits

weinig opened the pull request:
#### dc904eb4098bab9698ede556c0f2cc6b9955d9ab
<pre>
Introduce a CompactVariant for efficient variants of pointers and types smaller than 56 bits
<a href="https://bugs.webkit.org/show_bug.cgi?id=285099">https://bugs.webkit.org/show_bug.cgi?id=285099</a>

Reviewed by Darin Adler.

Adds `WTF::CompactVariant`.

A `WTF::CompactVariant` acts like a `std::variant` with the following differences:
- All alternatives must be pointers, smart pointers, have size of 56 bits or fewer,
  or be specialized for `CompactVariantTraits`.
- It can only contain 254 or fewer alternatives.
- Has a more limited API, only offering `holds_alternative()` for type checking and
  `switchOn()` for value access.

Factors out a few share helpers from VariantList into a VariantExtras.h file.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CompactVariant.h: Added.
* Source/WTF/wtf/CompactVariantOperations.h: Added.
* Source/WTF/wtf/VariantExtras.h: Added.
* Source/WTF/wtf/VariantList.h:
* Source/WTF/wtf/VariantListOperations.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp: Added.
* Tools/TestWebKitAPI/Tests/WTF/LifecycleLogger.h:

Canonical link: <a href="https://commits.webkit.org/288271@main">https://commits.webkit.org/288271@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca9a0b2ceaabf7f9d5f3378724fe8966e46e3af9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82335 "26 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36410 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87467 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33396 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84440 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9882 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/64169 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21920 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1457 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74919 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44447 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29100 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32437 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75323 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72649 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29730 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88823 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81389 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9641 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6900 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72570 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9867 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70733 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71787 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17887 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15922 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14931 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9594 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15115 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103802 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9468 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25193 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11238 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/CompactVariant.h:
```diff
@@ -0,0 +1,163 @@
+
+    CompactVariant& operator=(const CompactVariant& other)
+    {
+        if (*this == other)
```
```suggestion
        if (m_data == &other.m_data)
```

darinadler commented on Source/WTF/wtf/CompactVariant.h:
```diff
@@ -0,0 +1,163 @@
+        Operations::move(m_data, other.m_data);
+
+        // Set `other` to the "moved from" state.
+        other.m_data = UINT64_MAX;
```
Would be nicer to use C++ instead of C style `UINT64_MAX` and maybe also a named constant for this special value.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -0,0 +1,206 @@
+        });
+    }
+
+    static void destruct(uint64_t data)
```
Maybe use a named type like `Storage` instead of `uint64_t` each time?

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -0,0 +1,206 @@
+        uint64_t maskedData = value & payloadMask;
+
+        if constexpr (CompactVariantTraits<T>::hasAlternativeRepresentation) {
+            T dst = CompactVariantTraits<T>::decode(maskedData);
```
I’d prefer a word rather than `dst`.

weinig committed:
Introduce a CompactVariant for efficient variants of pointers and types smaller than 56 bits
https://bugs.webkit.org/show_bug.cgi?id=285099

Reviewed by Darin Adler.

Adds `WTF::CompactVariant`.

A `WTF::CompactVariant` acts like a `std::variant` with the following differences:
- All alternatives must be pointers, smart pointers, have size of 56 bits or fewer,
  or be specialized for `CompactVariantTraits`.
- It can only contain 254 or fewer alternatives.
- Has a more limited API, only offering `holds_alternative()` for type checking and
  `switchOn()` for value access.

Factors out a few share helpers from VariantList into a VariantExtras.h file.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CompactVariant.h: Added.
* Source/WTF/wtf/CompactVariantOperations.h: Added.
* Source/WTF/wtf/VariantExtras.h: Added.
* Source/WTF/wtf/VariantList.h:
* Source/WTF/wtf/VariantListOperations.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp: Added.
* Tools/TestWebKitAPI/Tests/WTF/LifecycleLogger.h:

Canonical link: https://commits.webkit.org/288271@main
